app/routes.py

```diff
@@ -1,2 +1,179 @@
+        db.session.commit()
+
+        return jsonify({"task":task.to_dict()}), 200
+
+@task_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_complete(task_id):
+    task = Task.query.get(task_id)
+
+    if task == None:
+        return make_response(), 404
+
+    task.completed_at = datetime.datetime.now()
+    db.session.commit()
+
+    url = "https://slack.com/api/chat.postMessage"
+    slack_headers = {'Authorization': "SLACK_BOT_TOKEN"}
+    slack_params = {'channel':'task-notifications',
+            'text': f'Someone just completed the task {task.description}',
+            'format': "json"}
+    requests.post(url, headers=slack_headers, params=slack_params)
+
+    return jsonify({"task":task.to_dict()}), 200
+
+@task_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_incomplete(task_id):
+    task = Task.query.get(task_id)
+
+    if task == None:
+        return make_response(), 404
+
+    task.completed_at = None
+    db.session.commit()
+
+    return jsonify({"task":task.to_dict()}), 200
+
+@goal_bp.route("", methods=["POST", "GET"])
+def goals():
+    if request.method == "POST":
+        request_body = request.get_json()
+        if "title" not in request_body:
+            return jsonify({"details": "Invalid data"}), 400
+        else:
+            new_goal = Goal(title = request_body["title"],)
+            db.session.add(new_goal)
+            db.session.commit()
+
+            return jsonify({"goal":new_goal.to_dict()}), 201
+
+    elif request.method == "GET":
+        goals = Goal.query.all()
+        goals_response = []
+        for goal in goals:
+            goals_response.append(goal.to_dict())
+
+        return jsonify(goals_response), 200
+
+@goal_bp.route("/<goal_id>", methods=["GET", "PUT", "DELETE"])
+def get_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+
+    if goal == None:
+        return make_response(), 404
+    
+    elif request.method == "GET":
+        return jsonify({"goal":goal.to_dict()}), 200
+
+    elif request.method == "DELETE":
+        db.session.delete(goal)
+        db.session.commit()
+        return jsonify({"details": f'Goal {goal.goal_id} "{goal.title}" successfully deleted'}),200
+
+    elif request.method == "PUT":
+        form_data = request.get_json()
+
+        goal.title = form_data["title"]
+        db.session.commit()
+
+        return jsonify({"goal":goal.to_dict()}), 200
+
+
+@goal_bp.route("/<goal_id>/tasks", methods=["GET", "POST"])
+def tasks_to_goals(goal_id):
+    goal = Goal.query.get(goal_id)
+
+    if goal == None:
+        return make_response(), 404
+    
+    elif request.method == "POST":
+        form_data = request.get_json() 
+        task_list = form_data["task_ids"]
+
+        for task in task_list:
+            task = Task.query.get(task)
+            task.assoc_goal = goal.goal_id
+            
+        db.session.commit()
+        return jsonify({"id": goal.goal_id,
+                        "task_ids": task_list}), 200
```

Comment:
This code looks great! This wasn't clarified in the requirements but it might be preferred behavior to return all tasks associated with the goal. This change would allow the response to include ALL tasks associated with the goal, not just the new ones.

```suggestion
        db.session.commit()
        full_task_list = []
        for task in goal.tasks:
            full_task_list.append(task.task_id)
        
        return jsonify({"id": goal.goal_id, "task_ids": full_task_list}), 200
```